### PR TITLE
Call import token if token not found

### PR DESCRIPTION
**Pull request by neikop**
https://team-kyber.atlassian.net/browse/KEP-1891

We didn't return token info when fetching a token that is not indexed.
-> Improve: Automatically call "import token" if query token not found.

### CODE REVIEW

File: src/hooks/Tokens.ts [hunk 1]
```diff
@@ -294,12 +294,22 @@ export const findCacheToken = (address: string) => {
   return cacheTokens[address] || cacheTokens[address.toLowerCase()]
 }
 
-export const fetchTokenByAddress = async (address: string, chainId: ChainId, signal?: AbortSignal) => {
+export const fetchTokenByAddress = async (address: string, chainId: ChainId, _signal?: AbortSignal) => {
```
Comment (viet-nv) [inline src/hooks/Tokens.ts:297]: remove the _signal if we dont need it

File: src/services/ksSetting.ts [hunk 2]
```diff
@@ -50,13 +50,22 @@ export type KyberswapGlobalConfigurationResponse = {
 
 export interface TokenListResponse<T = TokenInfo> {
   data: {
-    pageination: {
```
Comment (namgold) [inline src/services/ksSetting.ts:53]: này của ông :danh: à :bonk:

File: src/hooks/Tokens.ts [hunk 3]
```diff
@@ -294,12 +294,22 @@ export const findCacheToken = (address: string) => {
...
   if (address === ZERO_ADDRESS) return NativeCurrencies[chainId]
   const findToken = findCacheToken(address)
   if (findToken && findToken.chainId === chainId) return findToken
-  const response = await axios.get(`${KS_SETTING_API}/v1/tokens?query=${address}&chainIds=${chainId}`, { signal })
-  const token = response?.data?.data?.tokens?.[0]
+  const tokenListRes = await store.dispatch(
+    ksSettingApi.endpoints.getTokenList.initiate({ chainId, addresses: address }),
+  )
+  let token = tokenListRes.data?.data.tokens[0]
```
Comment (nhd98z) [inline src/hooks/Tokens.ts:304]: :+1

File: src/hooks/Tokens.ts [hunk 4]
```diff
@@ -294,12 +294,22 @@ export const findCacheToken = (address: string) => {
...
 }
 
-export const fetchTokenByAddress = async (address: string, chainId: ChainId, signal?: AbortSignal) => {
+export const fetchTokenByAddress = async (address: string, chainId: ChainId, _signal?: AbortSignal) => {
   if (address === ZERO_ADDRESS) return NativeCurrencies[chainId]
   const findToken = findCacheToken(address)
   if (findToken && findToken.chainId === chainId) return findToken
-  const response = await axios.get(`${KS_SETTING_API}/v1/tokens?query=${address}&chainIds=${chainId}`, { signal })
-  const token = response?.data?.data?.tokens?.[0]
+  const tokenListRes = await store.dispatch(
```
Comment (nguyenhoaidanh) [inline src/hooks/Tokens.ts:301]: do we need try catch?

File: src/services/ksSetting.ts (hunk 2 above)
Comment (nguyenhoaidanh) [reply src/services/ksSetting.ts:53]: 🐠

File: src/hooks/Tokens.ts [hunk 5]
```diff
@@ -294,12 +294,22 @@ export const findCacheToken = (address: string) => {
...
+    ksSettingApi.endpoints.getTokenList.initiate({ chainId, addresses: address }),
+  )
+  let token = tokenListRes.data?.data.tokens[0]
+  if (!token) {
+    const importTokenRes = await store.dispatch(
+      ksSettingApi.endpoints.importToken.initiate([{ chainId: chainId.toString(), address }]),
+    )
+    if ('data' in importTokenRes) {
+      token = importTokenRes.data.data.tokens[0]?.data
```
Comment (nhd98z) [inline src/hooks/Tokens.ts:310]: Might not need `?`.
```suggestion
      token = importTokenRes.data.data.tokens[0].data
```

Anyway I prefer `importTokenRes?.data?.data?.tokens?.[0]?.data` so no need `'data' in importTokenRes`.

Comment (nguyenhoaidanh) [reply src/hooks/Tokens.ts:310]:
```suggestion
      token = importTokenRes?.data?.data?.tokens?.[0]?.data
```

Comment (nguyenhoaidanh) [reply src/hooks/Tokens.ts:310]: or add transformResponse [here](https://github.com/KyberNetwork/kyberswap-interface/blob/5f08d1099d2e0fe16eaf82a30851d5f2d675a6dd/src/services/ksSetting.ts#L121) 

and then
```token = importTokenRes?.tokens?.[0]?.data```

**Comment by nguyenhoaidanh** [comment]
make sure QA tested it.

File: src/hooks/Tokens.ts [hunk 6]
```diff
@@ -294,12 +294,22 @@ export const findCacheToken = (address: string) => {
...
   if (findToken && findToken.chainId === chainId) return findToken
-  const response = await axios.get(`${KS_SETTING_API}/v1/tokens?query=${address}&chainIds=${chainId}`, { signal })
-  const token = response?.data?.data?.tokens?.[0]
+  const tokenListRes = await store.dispatch(
+    ksSettingApi.endpoints.getTokenList.initiate({ chainId, addresses: address }),
+  )
+  let token = tokenListRes.data?.data.tokens[0]
+  if (!token) {
+    const importTokenRes = await store.dispatch(
+      ksSettingApi.endpoints.importToken.initiate([{ chainId: chainId.toString(), address }]),
```
Comment (nguyenhoaidanh) [inline src/hooks/Tokens.ts:307]: do we need try catch?

File: src/hooks/Tokens.ts (hunk 5 above)
Comment (neikop) [reply src/hooks/Tokens.ts:310]: > Might not need `?`.

It needed, `tokens[0]` could be undefined


> Anyway I prefer `importTokenRes?.data?.data?.tokens?.[0]?.data` so no need `'data' in importTokenRes`.

It's an issue with typescript not checking undefined, `importTokenRes` has type kind of unknown

File: src/hooks/Tokens.ts (hunk 6 above)
Comment (neikop) [reply src/hooks/Tokens.ts:307]: I think we don't need try catch there

**Review by nguyenhoaidanh — DISMISSED** [review]

**Review by viet-nv — APPROVED** [review]

**Pull request merged**